Added paths filter to GitHub actions

iamdharmesh opened the pull request:
### Description of the Change
This PR adds a paths filter in GitHub action workflows to run it only on specific file extensions. For example, This PR doesn't contain any PHP, JS, or CSS file changes. So, the related workflows(E2E tests, PHPUnit, PHPCS, PHP Compatibility, and ESLint) will not be run against this PR. 

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #203 

### How to test the Change
See GitHub Actions check on this PR, E2E tests, PHPUnit, PHPCS, PHP Compatibility, and ESLint check should not run on this PR as we made changes in workflow files only.

### Changelog Entry
> Changed - Made GitHub Action workflow run only when required.

### Credits
Props @iamdharmesh

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

iamdharmesh committed:
Make E2E tests run only when PHP, JS or CSS files get changed

iamdharmesh committed:
Run eslint check only when JS files are changed

iamdharmesh committed:
Run PHP Compatibility check only when PHP files are changed

iamdharmesh committed:
Run PHPCS check only when PHP files are changed

iamdharmesh committed:
Run PHPUnit check only when PHP files are changed

peterwilsoncc reviewed (dismissed):
Fake approval to see what happens.

peterwilsoncc committed:
Merge branch 'develop' into enhancement/203

peterwilsoncc reviewed (dismissed):
Another fake approval to test things now that develop has been merged in.

peterwilsoncc reviewed (changes requested):
Sorry for the review spam while I was testing the effects of this PR.

In its current state, it prevents merging of approved PRs as the required actions have not completed.

I think what you'll want to do is something like this (pseudo code)

```yaml
step:
  Name: Succeed early if files are not modified
  run: |
    if ( no_relevant_file_modified() ) {
      finish_action_with_success();
    }
```

I'm not sure if that is possible though, it's something you'll need to check in the docs.

![Screen Shot 2022-12-01 at 1 41 46 pm](https://user-images.githubusercontent.com/519727/204953366-75f9dfea-56d5-46fc-948f-6c5aa4738679.png)

iamdharmesh committed:
Add generic PHPCS workflow.

iamdharmesh committed:
Merge branch 'enhancement/203' of github.com:10up/autoshare-for-twitter into enhancement/203

iamdharmesh commented:
@peterwilsoncc we can handle this by adding a generic workflow that ignores the path we added to the actual workflow So, it will run instead of the actual workflow and we can satisfy the `required` condition. More info can be found [here](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/troubleshooting-required-status-checks#handling-skipped-but-required-checks)

I have created and added that workflow for PHPCS([c4d298d](https://github.com/10up/autoshare-for-twitter/pull/204/commits/c4d298dade8f3527dbe9cfd9752544df38cdadfe)) and it works as expected. what do you think about this approach? If you think adding an additional dummy workflow for each required workflow is additional overhead then we might add path filters only on heavy workflows like E2E tests OR maybe adjust required workflow settings.

Thanks
cc: @dkotter

peterwilsoncc commented:
Thank @iamdharmesh -- that looks like a really nice approach. 

It does require repeating the paths in two files, so would need comments to alert people to change them in both locations. You could use a constant but I think that for two instances that would be over-engineering.

iamdharmesh committed:
Merge branch 'develop' into enhancement/203

iamdharmesh committed:
Added more dummy actions.

iamdharmesh committed:
Added matrix in ignore workflow.

peterwilsoncc commented on .github/workflows/ignore-eslint.yml:
```diff
@@ -0,0 +1,14 @@
+  eslint:
+    runs-on: ubuntu-latest
+    steps:
+      - run: 'echo "No ESLint check required"'
```
Nit: EOF line break would be lovely.

peterwilsoncc commented on .github/workflows/ignore-phpunit.yml:
```diff
@@ -0,0 +1,18 @@
+      matrix:
+        php-versions: ['7.4', '8.0']
+    steps:
+      - run: 'echo "No PHPUnit check required"'
```
```suggestion
      - run: 'echo "No PHPUnit check required"'

```

Nit: EOF line break would be lovely

peterwilsoncc reviewed (dismissed):
LGTM with a minor nitpick inline.

I tested the changes with #209 to ensure that the tests run and fail as expected when needed.

iamdharmesh committed:
Added EOF line break in .github/workflows/ignore-phpunit.yml

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

iamdharmesh committed:
Added EOF line break in ignore-eslint.yml

iamdharmesh committed:
Merge branch 'develop' into enhancement/203

peterwilsoncc reviewed (approved):
Thank you! LGTM.
